Increases dangers from SM delamination

atlantiscze opened the pull request:
A feedback thread will be open on the forums and linked to in this PR. Please keep all code-unrelated feedback to that thread.

EDIT: Thread located at: https://baystation12.net/forums/threads/feedback-supermatter-delamination-changes.2655/

Almost everything is in the form of defines, so it can be further tweaked easily.

DNM at least until 18.9.2016 - allowing time for feedback on the forums.
- Radiation in general is more dangerous, and has a third "stage" in which it deals 0.8 Tox damage per tick. This stage is only reached by long exposure or delamination. Radiation can also build up to 5 times larger number than it used to, making larger exposures actually highly dangerous and lethal without medical aid.
- Supermatter stabilization/destabilization is now faster, by roughly 50% (the core takes/loses damage faster). This is still (that was unchanged) influenced with the energy level. More emitter shots = faster destabilization.
- Ejected supermatter no longer has a special treatment for ejection that makes it miss the station in 99.9% of cases. That should make ejection a bit riskier. After all, it's a last resort thing, and should be expected to have it's risks.
- Delamination radiation and hallucinations are now 50% range based (linear scaling) and 50% Z level based. In other words, if you are on the same Z level you will always take at least half of the radiation and hallucinations.
- Delamination now weakens mobs for 20 ticks on explosion.
- Delamination now releases an electrical pulse. All APCs and SMESes on the same Z level will be affected and shut down, as if a grid check occured. This also affects critical APCs, but only for a fraction of the time. There is also a small (20%) RNG involved that adjusts the duration. Tested and shouldn't be lethal to AIs, assuming they don't panic and manage their power or manually reboot their SMES. In general, this is a combination of grid check and an electrical storm, even with the blown lights.
- Delamination now breaks majority of solar panels. It doesn't make the array nonfunctional completely - it only has a certain probability (60% for now) to break every individual solar panel on the same Z level. That should make it more threatening to station's power generation. These solars can be repaired relatively easily, but it requires some intervention from the engineers.
- Delamination explosion tweaked a bit. In general, it's strength is more or less the same, however it should have slightly lower devastation range, while having quite considerably larger light damage range (which carries a risk of damaging adjacent rooms slightly, it shouldn't breach those in 99% of cases however)
- Removes some outdated and commented out code relevant to radiation collectors.

atlantiscze committed:
Increases dangers from SM delamination
- Radiation in general is more dangerous, and has a third "stage" in which it deals 0.8 Tox damage per tick. This stage is only reached by long exposure or delamination. Radiation can also build up to 5 times larger number than it used to, making larger exposures actually highly dangerous and lethal without medical aid.
- Supermatter stabilization/destabilization is now faster, by roughly 50% (the core takes/loses damage faster). This is still (that was unchanged) influenced with the energy level. More emitter shots = faster destabilization.
- Ejected supermatter no longer has a special treatment for ejection that makes it miss the station in 99.9% of cases. That should make ejection a bit riskier. After all, it's a last resort thing, and should be expected to have it's risks.
- Delamination radiation and hallucinations are now 50% range based (linear scaling) and 50% Z level based. In other words, if you are on the same Z level you will always take at least half of the radiation and hallucinations.
- Delamination now weakens mobs for 20 ticks on explosion.
- Delamination now releases an electrical pulse. All APCs and SMESes on the same Z level will be affected and shut down, as if a grid check occured. This also affects critical APCs, but only for a fraction of the time. There is also a small (20%) RNG involved that adjusts the duration. Tested and shouldn't be lethal to AIs, assuming they don't panic and manage their power or manually reboot their SMES. In general, this is a combination of grid check and an electrical storm, even with the blown lights.
- Delamination now breaks majority of solar panels. It doesn't make the array nonfunctional completely - it only has a certain probability (60% for now) to break every individual solar panel on the same Z level. That should make it more threatening to station's power generation. These solars can be repaired relatively easily, but it requires some intervention from the engineers.
- Delamination explosion tweaked a bit. In general, it's strength is more or less the same, however it should have slightly lower devastation range, while having quite considerably larger light damage range (which carries a risk of damaging adjacent rooms slightly, it shouldn't breach those in 99% of cases however)
- Removes some outdated and commented out code relevant to radiation collectors.

PsiOmegaDelta commented on code/modules/supermatter/supermatter.dm:
```diff
@@ -20,13 +20,26 @@
+#define DETONATION_SHUTDOWN_CRITAPC 10		// Critical APC. AI core and such. Considerably shorter as we don't want to kill the AI with a single blast. Still a nuisance.
+#define DETONATION_SHUTDOWN_SMES 60			// SMES
+#define DETONATION_SHUTDOWN_RNG_FACTOR 20	// RNG factor. Above shutdown times can be +- X%, where this setting is the percent. Do not set to 100 or more.
+#define DETONATION_SOLAR_BREAK_CHANCE 60	// prob() of breaking solar arrays (this is per-panel, and only affects the Z level SM is on)
```
Should undef defines at the end of file unless used elsewhere. Reduces the risk of naming conflicts (or helps detect them) and apparently help a little with compilation times.

PsiOmegaDelta commented on code/modules/supermatter/supermatter.dm:
```diff
@@ -96,23 +109,67 @@
 	anchored = 1
 	grav_pulling = 1
 	exploded = 1
+	sleep(pull_time)
```
Avoid sleeping in procs called by `process()` unless you begin the proc with `set waitfor = 0` or you'll delay the processor.

PsiOmegaDelta commented on code/modules/supermatter/supermatter.dm:
```diff
@@ -96,23 +109,67 @@
+		if(TS && TM && (TS.z == TM.z))
+			var/range_multiplier = between(0, get_dist(mob, src) / world.maxx, 1)
+			var/rads = DETONATION_RADS_BASE + (range_multiplier * DETONATION_RADS_RANGE)
+			var/hallucinations = DETONATION_HALLUCINATION_BASE + (range_multiplier * DETONATION_HALLUCINATION_RANGE)
```
If memory serves hallucinations is reduced by 1 every 2 ticks. This basically means the minimum hallucination duration is 600 seconds (300 + (0* 200), while it was previously capped at 600.

PsiOmegaDelta commented on code/modules/supermatter/supermatter.dm:
```diff
@@ -96,23 +109,67 @@
+			var/range_multiplier = between(0, get_dist(mob, src) / world.maxx, 1)
+			var/rads = DETONATION_RADS_BASE + (range_multiplier * DETONATION_RADS_RANGE)
+			var/hallucinations = DETONATION_HALLUCINATION_BASE + (range_multiplier * DETONATION_HALLUCINATION_RANGE)
+			mob.apply_effect(rads, IRRADIATE)
```
Consider making this call `mob.apply_effect(rads, IRRADIATE, mob.getarmor(null, "rad"))`

Techhead0 replied:
Might be worthwhile to apply the rad protection to the hallucinations as well.

mustafakalash commented:
Coming back to this?

atlantiscze commented:
I plan to finish this PR (as well as the cyborg rebalance PR) once i am finished with the new shield generators. They're almost done, so i assume it will be either today or tommorow.

atlantiscze committed:
Implements feedback changes
- Implements changes requested by PsiOmegaDelta
- Reduces delamination mob weaken time considerably, as suggested on the forums.

PsiOmegaDelta commented on code/modules/supermatter/supermatter.dm:
```diff
@@ -96,23 +108,69 @@
-		explosion(get_turf(src), explosion_power, explosion_power * 1.25, explosion_power * 1.5, explosion_power * 1.75, 1)
+				H.hallucination += hallucinations
+			mob.Weaken(DETONATION_MOB_CONCUSSION)
+			mob << "<span class='danger'>An invisible force slams you against the ground!</span>"
```
to_chat(mob, msg)

PsiOmegaDelta reviewed (approved):
With no additional negative feedback since the latest changes any balance issues are left as future tweaking.
